drivers/watchdog/wdt_sam0.c

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 2017 Google LLC.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <soc.h>
+#include <watchdog.h>
+
+/* Device constant configuration parameters */
+struct wdt_sam0_dev_cfg {
+	Wdt *regs;
+};
+
+#define DEV_CFG(dev)							       \
+	((const struct wdt_sam0_dev_cfg *const)(dev)->config->config_info)
+
+static void wait_synchronization(Wdt *const wdt)
+{
+	while (wdt->STATUS.bit.SYNCBUSY) {
+	}
+}
+
+static void wdt_sam0_enable(struct device *dev)
+{
+	Wdt *const wdt = DEV_CFG(dev)->regs;
+
+	wdt->CTRL.reg = WDT_CTRL_ENABLE;
+	wait_synchronization(wdt);
+}
+
+static void wdt_sam0_disable(struct device *dev)
+{
+	Wdt *const wdt = DEV_CFG(dev)->regs;
+
+	wdt->CTRL.reg = 0;
+	wait_synchronization(wdt);
+}
+
+static int wdt_sam0_set_config(struct device *dev, struct wdt_config *config)
+{
+	return -ENOTSUP;
+}
+
+static void wdt_sam0_get_config(struct device *dev, struct wdt_config *config)
+{
+}
+
+static void wdt_sam0_reload(struct device *dev)
+{
+	Wdt *const wdt = DEV_CFG(dev)->regs;
+
+	wdt->CLEAR.bit.CLEAR = WDT_CLEAR_CLEAR_KEY_Val;
+}
+
+static const struct wdt_driver_api wdt_sam0_api = {
+	.enable = wdt_sam0_enable,
+	.disable = wdt_sam0_disable,
+	.get_config = wdt_sam0_get_config,
+	.set_config = wdt_sam0_set_config,
+	.reload = wdt_sam0_reload,
+};
+
+static int wdt_sam0_init(struct device *dev)
+{
+	Wdt *const wdt = DEV_CFG(dev)->regs;
+
+	/* Enable APB clock */
+	PM->APBAMASK.bit.WDT_ = 1;
+
+	/* Connect to GCLK2 (~1 kHz) */
+	GCLK->CLKCTRL.reg =
+	    GCLK_CLKCTRL_ID_WDT | GCLK_CLKCTRL_GEN_GCLK2 | GCLK_CLKCTRL_CLKEN;
+
+	wdt_sam0_disable(dev);
+
+	/* Pick the longest timeout */
+	wdt->CONFIG.reg = WDT_CONFIG_PER_16K;
+	wait_synchronization(wdt);
+
+#ifndef CONFIG_WDT_SAM0_DISABLE_AT_BOOT
+	wdt_sam0_enable(dev);
```

Comment:
I think it's better not to configure (enable) watchdog in the initialization function. Let's leave it to the user.

Reply:
I've re-worked and extended the watchdog driver.  The user can now do a:

wdt_enable()
wdt_config()

as per samples/drivers/watchdog.